// telegraf-inline-menu/dist/menu-responders.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const telegraf_1 = require("telegraf");
const middleware_helper_1 = require("./middleware-helper");
const combined_middleware_1 = __importDefault(require("./combined-middleware"));
class MenuResponders {
    constructor() {
        this.responders = [];
    }
    add(responder) {
        this.responders.push(responder);
    }
    hasSomeNonActionResponders() {
        return this.responders.some(o => o.action === undefined);
    }
    createMiddleware(environment) {
        const { actionCode, setMenuFunc } = environment;
        const menuMiddleware = telegraf_1.Composer.action(actionCode.get(), (ctx) => __awaiter(this, void 0, void 0, function* () { return setMenuFunc(ctx, 'menu action'); }));
        return telegraf_1.Composer.compose([
            menuMiddleware,
            ...this.responders
                .map(o => createMiddlewareFromResponder(o, environment))
        ]);
    }
}
exports.default = MenuResponders;
function createMiddlewareFromResponder(responder, environment) {
    const m = new combined_middleware_1.default(responder.middleware);
    if (responder.only) {
        m.addOnly(responder.only);
    }
    if (responder.hide) {
        m.addHide(responder.hide);
    }
    const { actionCode, setMenuFunc, setParentMenuFunc } = environment;
    if (responder.setParentMenuAfter) {
        if (!setParentMenuFunc) {
            throw new Error(`There is no parent menu for this that could be set. Remove the 'setParentMenuAfter' flag. Occured in menu ${actionCode.get()}`);
        }
        m.addAfterFunc((ctx) => __awaiter(this, void 0, void 0, function* () { return setParentMenuFunc(ctx, `setParentMenuAfter ${actionCode.get()}`); }), Boolean(responder.action));
    }
    else if (responder.setMenuAfter) {
        m.addAfterFunc((ctx) => __awaiter(this, void 0, void 0, function* () { return setMenuFunc(ctx, `setMenuAfter ${actionCode.get()}`); }), Boolean(responder.action));
    }
    if (!responder.action) {
        return m.middleware();
    }
    const childActionCode = actionCode.concat(responder.action);
    m.addOnly(middleware_helper_1.isCallbackQueryActionFunc(childActionCode));
    return m.middleware();
}
exports.createMiddlewareFromResponder = createMiddlewareFromResponder;
